les/vflux/server/clientpool.go

```diff
@@ -0,0 +1,375 @@
+// Stop shuts the client pool down. The clientPeer interface callbacks will not be called
+// after Stop. Register calls will return nil.
+func (cp *ClientPool) Stop() {
+	cp.balanceTracker.stop()
+	cp.ns.Stop()
+}
+
+// Register registers the peer into the client pool. If the peer has insufficient
+// priority and remains inactive for longer than the allowed timeout then it will be
+// disconnected by calling the Disconnect function of the clientPeer interface.
+func (cp *ClientPool) Register(peer clientPeer) ConnectedBalance {
+	cp.ns.SetField(peer.Node(), clientField, clientPeerInstance{peer})
+	balance, _ := cp.ns.GetField(peer.Node(), btSetup.balanceField).(*nodeBalance)
+	return balance
+}
+
+// Unregister removes the peer from the client pool
+func (cp *ClientPool) Unregister(peer clientPeer) {
+	cp.ns.SetField(peer.Node(), clientField, nil)
+}
+
+// SetDefaultFactors sets the default price factors applied to subsequently connected clients
+func (cp *ClientPool) SetDefaultFactors(posFactors, negFactors PriceFactors) {
+	cp.lock.Lock()
+	cp.defaultPosFactors = posFactors
+	cp.defaultNegFactors = negFactors
+	cp.lock.Unlock()
+}
+
+// setConnectedBias sets the connection bias, which is applied to already connected clients
+// So that already connected client won't be kicked out very soon and we can ensure all
+// connected clients can have enough time to request or sync some data.
+func (cp *ClientPool) SetConnectedBias(bias time.Duration) {
+	cp.lock.Lock()
+	cp.connectedBias = bias
+	cp.setActiveBias(bias)
+	cp.lock.Unlock()
+}
+
+// SetCapacity sets the assigned capacity of a connected client
+func (cp *ClientPool) SetCapacity(node *enode.Node, reqCap uint64, bias time.Duration, requested bool) (capacity uint64, err error) {
+	cp.lock.RLock()
+	if cp.connectedBias > bias {
+		bias = cp.connectedBias
+	}
+	cp.lock.RUnlock()
+
+	cp.ns.Operation(func() {
+		balance, _ := cp.ns.GetField(node, btSetup.balanceField).(*nodeBalance)
+		if balance == nil {
+			err = ErrNotConnected
+			return
+		}
+		capacity, _ = cp.ns.GetField(node, ppSetup.capacityField).(uint64)
+		if capacity == 0 {
+			// if the client is inactive then it has insufficient priority for the minimal capacity
+			// (will be activated automatically with minCap when possible)
+			return
+		}
+		if reqCap < cp.minCap {
+			// can't request less than minCap; switching between 0 (inactive state) and minCap is
+			// performed by the server automatically as soon as necessary/possible
+			reqCap = cp.minCap
+		}
+		if reqCap > cp.minCap {
+			if cp.ns.GetState(node).HasNone(btSetup.priorityFlag) && reqCap > cp.minCap {
+				err = ErrNoPriority
+				return
+			}
+		}
+		if reqCap == capacity {
+			return
+		}
+		curveBias := bias
+		if requested {
+			// mark the requested node so that the UpdateCapacity callback can signal
+			// whether the update is the direct result of a SetCapacity call on the given node
+			cp.capReqNode = node
+			defer func() {
+				cp.capReqNode = nil
+			}()
```

Comment:
Can we do the notification in this function(for the capacity updating)? It's unnecessary to rely on an external callback to this...

Reply:
I did think a bit about this and this is the best solution I have found so far. The reason I am using a callback here is because the capacity update can be initiated either from the client or server side. If the priority pool reduces the capacity then the server handler needs to be notified in a callback so that it can update its flow control params and send an update message to the client. If the client requests a capacity update and the server handler calls `SetCapacity` then basically the same thing should happen, the same callback will do the update job. The alternative would be to disable calling the callback when it was initiated with `SetCapacity` and leave it to the server handler to call `clientPeer.UpdateCapacity`. That would usually be fine but it is possible that a capacity granted by `SetCapacity` is instantly reduced by the priority pool (despite the bias that usually prevents this). In this case the order of events could be mixed up for the server handler, the callback resulting from the priority pool's reduction could be processed before the update made my the server handler after `SetCapacity` has returned. So I felt like this is the cleaner way.